security/advancedtls/advancedtls.go

```diff
@@ -79,21 +81,67 @@ type GetRootCAsResults struct {
 	TrustCerts *x509.CertPool
 }
 
-// RootCertificateOptions contains a field and a function for obtaining root
-// trust certificates.
-// It is used by both ClientOptions and ServerOptions.
-// If users want to use default verification, but did not provide a valid
-// RootCertificateOptions, we use the system default trust certificates.
+// RootCertificateOptions contains options to obtain root trust certificates
+// for both the client and the server.
+// At most one of the options could be set. If none of them are set, we
+// use the system default trust certificates.
 type RootCertificateOptions struct {
-	// If field RootCACerts is set, field GetRootCAs will be ignored. RootCACerts
-	// will be used every time when verifying the peer certificates, without
-	// performing root certificate reloading.
+	// If RootCACerts is set, it will be used every time when verifying
+	// the peer certificates, without performing root certificate reloading.
 	RootCACerts *x509.CertPool
-	// If GetRootCAs is set and RootCACerts is nil, GetRootCAs will be invoked
-	// every time asked to check certificates sent from the server when a new
-	// connection is established.
-	// This is known as root CA certificate reloading.
-	GetRootCAs func(params *GetRootCAsParams) (*GetRootCAsResults, error)
+	// If GetRootCertificates is set, it will be invoked to obtain root certs in
+	// every new connection.
+	GetRootCertificates func(params *GetRootCAsParams) (*GetRootCAsResults, error)
+	// If RootProvider is set, we will use the root certs from the Provider's
+	// KeyMaterial() call in every new connection. The Provider must have initial
+	// credentials if specified. Otherwise, KeyMaterial() will block forever.
+	RootProvider certprovider.Provider
+}
+
+// nonNilFieldCount returns the number of set fields in RootCertificateOptions.
+func (o RootCertificateOptions) nonNilFieldCount() int {
+	cnt := 0
+	rv := reflect.ValueOf(o)
+	for i := 0; i < rv.NumField(); i++ {
+		if !rv.Field(i).IsNil() {
+			cnt++
+		}
+	}
+	return cnt
+}
+
+// IdentityCertificateOptions contains options to obtain identity certificates
+// for both the client and the server.
+// At most one of the options could be set.
+type IdentityCertificateOptions struct {
```

Comment:
`IdentityCertificate` seems a little weird to me. How about `LeafCertificate` or `CertificateChain` depending on whether it contains intermediate certificates or not?

Reply:
Or just using `Certificate`? Since it's sufficient to distinguish it from `RootCertificate`.

Reply:
I first saw the name "identity certificate" couple of times in some gRPC team's proposals, and I soon started to use it,  because I think it conveys the usage of this certificate(which is to present identity to the peers), just like "root certificate"(which represents root of trust). Now it is used in many places in C core...
From this perspective, a word like `Certificate` doesn't serve this purpose, because technically speaking, a root certificate is also a certificate. Can we keep the word if you don't mind? :)

Reply:
Ah Ok. If `identity certificate` is an agreed upon term within gRPC, I am fine with it.